src/index.rs

```diff
@@ -412,6 +412,191 @@ impl Index {
         unsafe { raw::git_index_has_conflicts(self.raw) == 1 }
     }
 
+    /// Add or update index entries to represent a conflict. Any staged entries
+    /// that exist at the given paths will be removed.
+    ///
+    /// The entries are the entries from the tree included in the merge. Any entry
+    /// may be `None` to indicate that that file was not present in the trees during
+    /// the merge. For example, ancestor_entry may be `None` to indicate that a file
+    /// was added in both branches and must be resolved.
+    pub fn conflict_add(
```

Comment:
The general intention of this crate is to be a lightweight binding over libgit2 itself, but this is a pretty heavyweight method. Would it be possible to avoid the movement around of all the data internally?

Reply:
I based this implementation on all the other methods taking a `IndexEntry` (e.g. [`Index::add()`](https://github.com/rust-lang/git2-rs/blob/master/src/index.rs#L147). This one is particularly long because it takes three `IndexEntry` parameters. I don't know if the handling of `IndexEntry`s can be more concise, my intention was more to handle them exactly the same as in the existing methods.

Reply:
Maybe [`std::mem::transmute`](https://doc.rust-lang.org/stable/std/mem/fn.transmute.html) can be used here, but I have no clue if `git_index_entry` is even remotely binary compatible with an `IndexEntry` especially regarding the `path: Vec<u8>`. Also, this would be incredibly easy to mess up and hard to debug :see_no_evil: 

Maybe someone has a better idea :smile:

Reply:
Ah ok, well in that case can the conversion into a `git_index_entry` be refactored into a helper function instead of being duplicated?

Reply:
Sure, will do so. Should I also refactor all conversions in other methods using `IndexEntry`?

Reply:
I guess? Sort of depends, I'm not intimately familiar with all this code.

Reply:
Alright, my solution is to use a helper function which supplies pointers to `raw::git_index_entry` instances via a callback. This is needed as the construction of a `git_index_entry` requires the construction of a `CString` to which the `git_index_entry` points to. I went for a callback solution as it is (in my opinion) the easiest approach to make sure that the `CString` is available while using the raw `git_index_entry`. The procedure is implemented in [`try_raw_entries()@src/index.rs:92`](https://github.com/fin-ger/git2-rs/blob/add-index-conflict-fns/src/index.rs#L92). The implementation is a bit messy and requires the construction of three `Vec`s. Suggestions welcome!